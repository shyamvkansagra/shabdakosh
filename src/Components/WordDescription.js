import React, { Component } from 'react';
import './style.css';

export default class SearchBox extends Component {
    render() {
        const { dictionaryWord, dictionaryDefinition } = this.props;
        return (
            <div>
                <div className="definitionContainer">
                    <span className="dictionaryWord">{dictionaryWord}</span>
                    <p className="dictionaryDefinition">{dictionaryDefinition}</p>
                </div>
            </div>
        );
    }
}