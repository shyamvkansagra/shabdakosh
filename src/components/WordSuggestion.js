import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './style.css';

export default class WordSuggestions extends Component {
	static propTypes = {
		suggestedWord: PropTypes.string.isRequired,
		searchSuggestedWord: PropTypes.func.isRequired,
	};

    render() {
        const { suggestedWord, searchSuggestedWord } = this.props;
        return (
            <button className="suggestedWord" type="button" data-word={suggestedWord} onClick={searchSuggestedWord}>
                {suggestedWord}
            </button>
        );
    }
}