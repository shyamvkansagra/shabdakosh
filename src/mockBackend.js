import levenshtein from 'liblevenshtein';
import DataStore from 'nedb';

const dictionaryTxtToObj = data => {
    const dataArr = data.split('\n\n');
    const dictionaryObj = {};
    dataArr.forEach(d => {
        const dataDefinition = d.split(' ');
        const [dictionaryWord] = dataDefinition;
        dataDefinition.shift();
        const dictionaryDefinition = dataDefinition.join(' ');
        dictionaryObj[dictionaryWord.toLowerCase()] = { dictionaryWord, dictionaryDefinition };
    });
   return dictionaryObj;
}

const fetchDictionary = () => fetch('https://raw.githubusercontent.com/shyamvkansagra/shabdakosh/development/dictionary.txt')
    .then(res => res.text())
    .then(data => data)
    .catch(err => console.log('Fetch Error:', err));

const buildSuggester = (dictionaryObj) => {
    const completionList = Object.keys(dictionaryObj);
    const builder = new levenshtein.Builder()
        .dictionary(completionList, false)  // generate spelling candidates from unsorted completion_list
        .algorithm("transposition")          // use Levenshtein distance extended with transposition
        .sort_candidates(true)               // sort the spelling candidates before returning them
        .case_insensitive_sort(true)         // ignore character-casing while sorting terms
        .include_distance(false)             // just return the ordered terms (drop the distances)
        .maximum_candidates(3);

    return builder.build();
}

const setDB = async (dictionaryObj) => {
    const db = new DataStore();
    db.ensureIndex({ fieldName: 'word' });
    
    const docsToBeInserted = [];
    Object
        .keys(dictionaryObj)
        .forEach(w => docsToBeInserted.push({
            word: w,
            dictionaryWord: dictionaryObj[w].dictionaryWord,
            dictionaryDefinition: dictionaryObj[w].dictionaryDefinition,
        }))

    await db.insert(docsToBeInserted);
    return db;
}

const mockBackend = async () => {
    const dictionaryTxt = await fetchDictionary();
    const dictionaryObj = dictionaryTxtToObj(dictionaryTxt);

    const suggester = buildSuggester(dictionaryObj);
    const db = await setDB(dictionaryObj);

    return { db, suggester };
}

export default mockBackend;



