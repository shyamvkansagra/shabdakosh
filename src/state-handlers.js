export const searchHandler = (searchTerm, newSearchTerm) => {
	if (searchTerm && !newSearchTerm) {
		return {
			searchTerm: '',
            wordSuggestions: [],
            noneFound: false,
		};
	}

	return { searchTerm: newSearchTerm };
}