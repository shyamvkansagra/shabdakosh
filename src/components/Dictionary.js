import React, { Component } from 'react';
import levenshtein from 'liblevenshtein';

// Import styles
import './style.css';

// Import components
import SearchBox from './SearchBox';
import WordDescription from './WordDescription';
import WordSuggestion from './WordSuggestion';

export default class Dictionary extends Component {
    state = {
        searchTerm: '',
        dictionaryWord: '',
        dictionaryDefinition: '',
        wordSuggestions: [],
        noneFound: false,
    };

    handleSearchChange = e => {
        if (this.state.searchTerm && !e.target.value) {
            // search term was reset
            this.setState({
                searchTerm: '',
                wordSuggestions: [],
                noneFound: false,
            })
        } else {
            // user is entering the data
            this.setState({ searchTerm: e.target.value });
        }
    }

    handleSearch = () => {
        const { searchTerm } = this.state;
        const { db, suggester } = this.props;
        
        db.findOne(
            { word: searchTerm.toLowerCase() },
            (err, doc) => {
                if (doc) {
                    return this.setState({
                        dictionaryWord: doc.dictionaryWord,
                        dictionaryDefinition: doc.dictionaryDefinition,
                        wordSuggestions: [],
                        noneFound: false,
                    });
                }

                const MAX_EDIT_DISTANCE = 2;
                const wordSuggestions = suggester.transduce(searchTerm, MAX_EDIT_DISTANCE);

                return this.setState({
                    wordSuggestions,
                    dictionaryWord: '',
                    dictionaryDefinition: '',
                    noneFound: wordSuggestions.length === 0,
                })
            }
        );
    }

    searchSuggestedWord = e => {
        const { word } = e.currentTarget.dataset;
        this.setState({ searchTerm: word, wordSuggestions: [], noneFound: false }, this.handleSearch);
    }

    render() {
        const {
            searchTerm,
            dictionaryWord,
            dictionaryDefinition,
            wordSuggestions,
            noneFound
        } = this.state;
        const showSuggestions = !dictionaryWord && !dictionaryDefinition && wordSuggestions;

        return (
            <div>
                <div className="searchContainer">
                    <span className="labelMsg">
                        <p>Know the meaning of that cool word you came across:</p>
                    </span>
                    <SearchBox
                        searchTerm={searchTerm}
                        handleSearchChange={this.handleSearchChange}
                        handleSearch={this.handleSearch}
                    />
                    {showSuggestions && showSuggestions.length
                    ? (
                        <div className="wordSuggestions">
                            <span>Did you mean?:</span>
                            {wordSuggestions.map(ws => (
                                <WordSuggestion
                                    key={ws}
                                    suggestedWord={ws}
                                    searchSuggestedWord={this.searchSuggestedWord}
                                />
                            ))}
                            <p>(Click any above suggested word for instant search)</p>
                        </div>
                    )
                    : (noneFound
                        ? (
                            <div className="noneFoundMsg">
                                <span>Nothing found! ðŸ™Š says, "Hmm, maybe this Shabdakosh is not as cool as I thought!" </span>
                            </div>
                        ) : null
                    )
                    }
                </div>
                {dictionaryWord && dictionaryDefinition &&
                    <WordDescription
                        dictionaryWord={dictionaryWord}
                        dictionaryDefinition={dictionaryDefinition}
                    />
                }
            </div>
        );
    }
}