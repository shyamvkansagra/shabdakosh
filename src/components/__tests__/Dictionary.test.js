import { shallow } from 'enzyme';
import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

// Import handler functions
import { searchHandler } from '../../state-handlers';

import mockBackend from '../../mockBackend';

// Import components
import Dictionary from '../Dictionary';
import WordSuggestion from '../WordSuggestion';
import WordDescription from '../WordDescription';

test('Search box should clear state if searched term is removed', () => {
	const searchTerm = 'something';
	const newSearchTerm = '';
	// Scenario where search term was cleared
	
	const returnData = searchHandler(searchTerm, newSearchTerm);

	expect(returnData).toEqual({
		searchTerm: '',
        wordSuggestions: [],
        noneFound: false,
	});
});

test('Dictionary component renders correctly', () => {
  const component = shallow(<Dictionary suggester={{}} db={{}} />);
  
  expect(component).toMatchSnapshot();
});

test('Clicking suggested word should call searchSuggestedWord method with that word', () => {
  const suggestedWord = "something";
  const searchSuggestedWord = jest.fn();
  const wrapper = mount(
    <WordSuggestion suggestedWord={suggestedWord} searchSuggestedWord={searchSuggestedWord} />
  );

  const btn = wrapper.find('.suggestedWord');
  btn.simulate('click');
  
  expect(searchSuggestedWord).toHaveBeenCalled();
});

test('WordDescription component should have proper word', () => {
  const dictionaryWord = "Something";
  const wrapper = mount(
    <WordDescription dictionaryWord={dictionaryWord} dictionaryDefinition="" />
  );
  const span = wrapper.find('.dictionaryWord');
  
  expect(span.text()).toBe('Something');
});

test('Searched word, if found, should fetch correct word', async () => {
	const { db } = await mockBackend();
	db.findOne({ word: "hi" }, (err, doc) => {
		expect(doc.dictionaryWord).toBe('Hi');
	});
});