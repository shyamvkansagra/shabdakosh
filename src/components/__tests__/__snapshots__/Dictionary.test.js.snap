// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dictionary
    db={Object {}}
    suggester={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="searchContainer"
        >
          <span
            className="labelMsg"
          >
            <p>
              Know the meaning of that cool word you came across:
            </p>
          </span>
          <WordDescription
            handleSearch={[Function]}
            handleSearchChange={[Function]}
            searchTerm=""
          />
        </div>,
        "",
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="labelMsg"
            >
              <p>
                Know the meaning of that cool word you came across:
              </p>
            </span>,
            <WordDescription
              handleSearch={[Function]}
              handleSearchChange={[Function]}
              searchTerm=""
            />,
            null,
          ],
          "className": "searchContainer",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <p>
                Know the meaning of that cool word you came across:
              </p>,
              "className": "labelMsg",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Know the meaning of that cool word you came across:",
              },
              "ref": null,
              "rendered": "Know the meaning of that cool word you came across:",
              "type": "p",
            },
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "handleSearch": [Function],
              "handleSearchChange": [Function],
              "searchTerm": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
        ],
        "type": "div",
      },
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="searchContainer"
          >
            <span
              className="labelMsg"
            >
              <p>
                Know the meaning of that cool word you came across:
              </p>
            </span>
            <WordDescription
              handleSearch={[Function]}
              handleSearchChange={[Function]}
              searchTerm=""
            />
          </div>,
          "",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="labelMsg"
              >
                <p>
                  Know the meaning of that cool word you came across:
                </p>
              </span>,
              <WordDescription
                handleSearch={[Function]}
                handleSearchChange={[Function]}
                searchTerm=""
              />,
              null,
            ],
            "className": "searchContainer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <p>
                  Know the meaning of that cool word you came across:
                </p>,
                "className": "labelMsg",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Know the meaning of that cool word you came across:",
                },
                "ref": null,
                "rendered": "Know the meaning of that cool word you came across:",
                "type": "p",
              },
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "handleSearch": [Function],
                "handleSearchChange": [Function],
                "searchTerm": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            null,
          ],
          "type": "div",
        },
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
